File: GreekReader/app/src/main/java/com/ericmschmidt/greekreader/utilities/DictionaryXMLHelper.java
package com.ericmschmidt.greekreader.utilities;

import com.ericmschmidt.greekreader.MyApplication;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import java.io.*;
import java.util.ArrayList;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import org.w3c.dom.*;

/**
 * Gets dictionary resources out of the XML.
 */
public class DictionaryXMLHelper extends XmlParserHelper {

    // Overrides constants in super class.
    private static final String sectionTag = "div0";
    private static final String lineTag = "entry";
    private static final String keyAttribute = "key";

    /*
    The XML dictionary is structured like this:

    <work> <!-- This corresponds to the Work data structure -->
      <header>
        <!-- Stuff here ... -->
      <header/>
      <text>
        <body> <!-- This is like a collection of Book objects -->
          <div0 type="alphabetic letter" n="A"> <!-- This is a alphabetic section in the dictionary.
          <head></head>
            <superEntry>
              <entry></entry> <!-- This is usually the entry for the letter itself. -->
              <entry></entry>
            </superEntry>
            <entry id="n2" type="main" key="abactus"> <!-- This is the actual dictionary entry for a word. -->
              <form><orth extent="full" lang="la">abctus</orth></form>
              <sense id="n2.0" level="0" n="0"><etym lang="la">P. of abigo</etym>,
                <trans><tr>driven away, driven off</tr></trans>:
                <foreign lang="la">nox abacta</foreign>,
                <trans><tr>driven back</tr></trans> (from the pole), i. e.
                <trans><tr>already turned towards dawn</tr></trans>,
                <usg>V.</usg>:
                <foreign lang="la">abacta null conscienti</foreign>,
                <trans><tr>restrained by</tr></trans>,
                <usg>H.</usg>
              </sense>
            </entry>
          </div0>
        </body>
      </text>
    </work>
    */

    /**
     * Gets the number of alphabetical sections in the dictionary.
     * @param stream the inputStream of the dictionary.
     * @return the count of the alphabetical sections.
     * @throws XmlPullParserException
     * @throws IOException
     */
    public static int getSectionCount(InputStream stream) throws XmlPullParserException, IOException {
        int count = 0;
        try {
            XmlPullParser parser = initParser(stream);
            count = countSections(parser, sectionTag);

        } catch (Exception ex) {
            String errorMessage = ex.getMessage();
            MyApplication.logError(errorMessage);
        } finally {
            stream.close();
        }
        return count;
    }

    /**
     * Gets the keys of all the entries in this dictionary.
     * @param stream the inputStream of the dictionary.
     * @return the ArrayList of dictionary entries.
     * @throws XmlPullParserException
     * @throws IOException
     */
    public static ArrayList<String> getEntryHeaders(InputStream stream) throws XmlPullParserException, IOException {
        ArrayList<String> headers = new ArrayList<String>();

        try {
            XmlPullParser parser = initParser(stream);

            while (nextSection(parser, lineTag)) {
                parser.require(XmlPullParser.START_TAG, XmlParserHelper.ns, lineTag);

                String keyAttributeValue = parser.getAttributeValue(XmlParserHelper.ns, keyAttribute);
                if (keyAttributeValue != null) {
                    headers.add(keyAttributeValue);
                }
            }

        } catch (Exception ex) {
            String errorMessage = ex.getMessage();
            MyApplication.logError(errorMessage);
        } finally {
            stream.close();
        }
        return headers;
    }
    /**
     * Gets the keys of all the entries in this dictionary from an entries resource.
     * @param stream the inputStream of the dictionary.
     * @return the ArrayList of dictionary entries.
     * @throws Exception
     */
    public static ArrayList<String> getEntryHeaders2(InputStream stream) throws Exception {
        ArrayList<String> headers = new ArrayList<String>();
        try {

            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            DocumentBuilder builder = factory.newDocumentBuilder();

            Document dom = builder.parse(stream);
            NodeList list = dom.getElementsByTagName("entry");

            for (int i = 0; i < list.getLength(); i++) {
                headers.add(list.item(i).getTextContent());
            }

        } catch (Exception ex) {
            String errorMessage = ex.getMessage();
            MyApplication.logError(errorMessage);
        } finally {
            stream.close();
        }
        return headers;
    }

    /**
     * Get a specific dictionary entry from the XML.
     * @param stream the inputStream of the dictionary.
     * @param searchEntry the dictionary entry to search for.
     * @return the definition for the entry.
     * @throws XmlPullParserException
     * @throws IOException
     */
    public static String getEntry(InputStream stream, String searchEntry) throws XmlPullParserException, IOException {
        String definition = null;

        try {
            XmlPullParser parser = initParser(stream);
            while (nextSection(parser, lineTag)) { // TODO: Generalize this loop for this and previous method.
                parser.require(XmlPullParser.START_TAG, XmlParserHelper.ns, lineTag);

                String keyAttributeValue = parser.getAttributeValue(XmlParserHelper.ns, keyAttribute);
                if (keyAttributeValue != null
                        && searchEntry.equals(keyAttributeValue)) {
                    break;
                }
            }

            definition = getLine(parser, lineTag);

        } catch (Exception ex) {
            String errorMessage = ex.getMessage();
            MyApplication.logError(errorMessage);
        } finally {
            stream.close();
        }

        return removeExtraneousCharacters(definition);
    }
}

File: LatinReader/core-data/src/main/java/com/telpirion/classicsreader/core_data/utilities/DictionaryXMLHelper.kt
package com.telpirion.classicsreader.core_data.utilities

import android.util.Log
import org.w3c.dom.Document
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.io.InputStream
import javax.xml.parsers.DocumentBuilderFactory

/**
 * Gets dictionary resources out of the XML.
 *
 * @author Eric Schmidt
 * @author <a href="http://telpirion.com">...</a>
 * @version 1.5
 * @since 1.0
 */
object DictionaryXMLHelper : XmlParserHelper() {

    private const val LINE_TAG = "entry"
    private const val KEY_ATTRIBUTE = "key"
    private const val LANG_ATTRIBUTE = "lang"

    /**
     * Gets the keys of all the entries in this dictionary from an entries resource.
     * @param stream the inputStream of the dictionary.
     * @return the ArrayList of dictionary entries.
     */
    @Throws(Exception::class)
    fun getEntryHeaders(stream: InputStream): ArrayList<String> {
        val headers = ArrayList<String>()
        try {
            val factory = DocumentBuilderFactory.newInstance()
            val builder = factory.newDocumentBuilder()

            val dom: Document = builder.parse(stream)
            val list = dom.getElementsByTagName("entry")

            for (i in 0 until list.length) {
                headers.add(list.item(i).textContent)
            }
        } catch (ex: Exception) {
            Log.e(this.javaClass.name, ex.message as String)
        } finally {
            stream.close()
        }
        return headers
    }

    /**
     * Get a specific dictionary entry from the XML.
     * @param stream the inputStream of the dictionary.
     * @param searchEntry the dictionary entry to search for.
     * @return the definition for the entry.
     */
    @Throws(XmlPullParserException::class, IOException::class)
    fun getEntry(stream: InputStream, searchEntry: String, converter: ITextConverter?): String? {
        var definition: String? = null

        try {
            val parser = initParser(stream)
            while (nextSection(parser, LINE_TAG)) {
                parser.require(XmlPullParser.START_TAG, ns, LINE_TAG)

                val keyAttributeValue = parser.getAttributeValue(ns, KEY_ATTRIBUTE)
                if (keyAttributeValue != null && searchEntry == keyAttributeValue) {
                    break
                }
            }

            definition = getDictionaryLine(parser, LINE_TAG, converter)

        } catch (ex: Exception) {
            Log.e(this.javaClass.name, ex.message as String)
        } finally {
            stream.close()
        }

        return definition?.let { removeExtraneousCharacters(it) }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun getDictionaryLine(parser: XmlPullParser,
                                  lineTag: String,
                                  converter: ITextConverter?): String {
        var line = ""
        var currentSubLine: String
        var isNonLatin = false
        parser.require(XmlPullParser.START_TAG, ns, lineTag)

        while (!checkTag(lineTag, parser, XmlPullParser.END_TAG)) {
            parser.next()

            if (parser.eventType == XmlPullParser.START_TAG && converter != null && hasAttribute(parser, LANG_ATTRIBUTE, converter.lang)) {
                isNonLatin = true
            }

            if (parser.eventType == XmlPullParser.TEXT) {
                currentSubLine = if (isNonLatin && converter != null) {
                    converter.convertSourceToTargetCharacters(parser.text)
                } else {
                    parser.text
                }
                line += removeExtraneousCharacters(currentSubLine)
                isNonLatin = false
            }
        }

        return line
    }

    private fun hasAttribute(parser: XmlPullParser, attributeName: String, languageName: String): Boolean {
        val attributeValue = parser.getAttributeValue(null, attributeName)
        return attributeValue != null && attributeValue == languageName
    }
}